cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCorePeerIdentityHandler.java

```diff
@@ -0,0 +1,126 @@
+ *    Rikard Höglund (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.elements.AddressEndpointContext;
+import org.eclipse.californium.elements.EndpointContext;
+import org.eclipse.californium.elements.auth.OSCorePeerIdentity;
+
+/**
+ * Class that handles the EndpointContext for OSCORE messages. 
+ *
+ * It provides functionality to set an appropriate OSCORE peer identity
+ * for an EndpointContext.
+ *
+ * Will only add a peer identity for endpoint contexts that have
+ * an empty or OSCORE peer identity. If for instance Scandium has
+ * set another type of Principal as peer identity it will not be updated.
+ *
+ */
+public class OSCorePeerIdentityHandler {
+
+	/**
+	 * Sets destination endpoint context for outgoing requests with an OSCORE peer identity.
+	 *
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param request the request to set the endpoint context for
+	 */
+	public static void sendingRequest(OSCoreCtx oscoreCtx, Request request) {
+		setPeerIdentityOutgoing(oscoreCtx, request);
+	}
+
+	/**
+	 * Sets source endpoint context for incoming requests with an OSCORE peer identity.
+	 *
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param request the request to set the endpoint context for
+	 */
+	public static void receivingRequest(OSCoreCtx oscoreCtx, Request request) {
+		setPeerIdentityIncoming(oscoreCtx, request);
+	}
+
+	/**
+	 * Sets destination endpoint context for outgoing responses with an OSCORE peer identity.
+	 *
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param response the response to set the endpoint context for
+	 */
+	public static void sendingResponse(OSCoreCtx oscoreCtx, Response response) {
+		setPeerIdentityOutgoing(oscoreCtx, response);
+	}
+
+	/**
+	 * Sets destination endpoint context for incoming responses with an OSCORE peer identity.
+	 *
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param response the response to set the endpoint context for
+	 */
+	public static void receivingResponse(OSCoreCtx oscoreCtx, Response response) {
+		setPeerIdentityIncoming(oscoreCtx, response);
+	}
+
+	/**
+	 * Updates the currently set destination endpoint context for outgoing
+	 * messages adding an OSCORE peer identity.
+	 *
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param message the message to set the destination endpoint context for
+	 */
+	private static void setPeerIdentityOutgoing(OSCoreCtx oscoreCtx, Message message) {
+		//Create a new OSCORE peer identity
+		OSCorePeerIdentity newPeerIdentity = new OSCorePeerIdentity(oscoreCtx.getUri(),
+				oscoreCtx.getIdContext(), oscoreCtx.getRecipientId(), oscoreCtx.getSenderId());
+
```

Comment:
`OSCorePeerIdentity(String host, byte[] contextId, byte[] senderId, byte[] recipientId)` sender and recipient id seems to be provided in exchanged order?

Reply:
Yes, I did this on purpose with the intention that the peer identity contains the settings of that peer. In other words the sender ID of the peer is my recipient ID, and the recipient ID of the peer is my sender ID. That's why the parameters taken from the local context are switched.